File: samples/imagePyramid_lua/src/ImagePyramid.cpp

#include "ImagePyramid.h"

#include <stb_image_write.h>

#include <array>
#include <iostream>


ImagePyramid::ImagePyramid(const std::shared_ptr<ll::Session>& session) :
    ll::ContainerNode(session) {
}


void ImagePyramid::record(ll::CommandBuffer& commandBuffer) const {

    for (auto h = 0u; h < 4; ++h) {
        auto downX = getNode("downX_" + std::to_string(h));
        auto downY = getNode("downY_" + std::to_string(h));

        downX->record(commandBuffer);
        commandBuffer.memoryBarrier();

        downY->record(commandBuffer);
        commandBuffer.memoryBarrier();
    }
}


void ImagePyramid::onInit() {

    auto in_RGBA = getPort("in_RGBA");

    // insert in_RGBA again as pyramid_0
    bind("pyramid_0", in_RGBA);

    std::cout << "changing layout" << std::endl;
    std::static_pointer_cast<ll::ImageView>(in_RGBA)->changeImageLayout(vk::ImageLayout::eGeneral);

    for (auto h = 0u; h < 4; ++h) {

        std::cout << "CPP: level: " << h  << std::endl;

        auto downX = m_session->createComputeNode("imageDownsampleX");
        auto downY = m_session->createComputeNode("imageDownsampleY");
        
        downX->bind("in_RGBA", in_RGBA);
        downX->init();
        
        downY->bind("in_RGBA", downX->getPort("out_RGBA"));
        downY->init();

        // the input to the next level will be the output of downY
        in_RGBA = downY->getPort("out_RGBA");
        bind("pyramid_" + std::to_string(h + 1), in_RGBA);

        bindNode("downX_" + std::to_string(h), downX);
        bindNode("downY_" + std::to_string(h), downY);
    }
}


void ImagePyramid::writeAllImages(std::shared_ptr<ll::Session> session) {

    constexpr const auto names = std::array<const char*, 4> {{"pyramid_0", "pyramid_1", "pyramid_2", "pyramid_3"}};

    for (const auto& name : names) {
        writeImage(session,
                   std::static_pointer_cast<ll::ImageView>(getPort(name))->getImage(),
                   std::string{name} + std::string{".bmp"});
    }
}


void ImagePyramid::writeImage(std::shared_ptr<ll::Session> session, std::shared_ptr<ll::Image> image, const std::string& filename) {

    const auto imageSize     = image->getSize();
    const auto currentLayout = image->getLayout();

    // create host visible memory
    const auto hostMemFlags = vk::MemoryPropertyFlagBits::eHostVisible | vk::MemoryPropertyFlagBits::eHostCoherent;
    auto hostMemory         = session->createMemory(hostMemFlags, imageSize);
    auto hostImage          = hostMemory->createBuffer(imageSize);

    auto cmdBuffer = session->createCommandBuffer();

    cmdBuffer->begin();
    cmdBuffer->changeImageLayout(*image, vk::ImageLayout::eTransferSrcOptimal);
    cmdBuffer->copyImageToBuffer(*image, *hostImage);
    cmdBuffer->changeImageLayout(*image, currentLayout);
    cmdBuffer->end();

    session->run(*cmdBuffer);

    auto mapPtr = hostImage->map<uint8_t>();
    const auto res = stbi_write_bmp(filename.c_str(),
        image->getWidth(),
        image->getHeight(),
        image->getChannelCount<uint32_t>(),
        mapPtr.get());

    std::cout << "stbi_write_bmp result: " << res << std::endl;
}

File: samples/imagePyramid_lua/lua/ImagePyramid.lua
local builder = ll.class(ll.ContainerNodeBuilder)


function builder.newDescriptor()

    ll.logd('ImagePyramid', 'newDescriptor')
    local desc = ll.ContainerNodeDescriptor.new()

    desc.builderName = 'ImagePyramid'

    desc:addPort(ll.PortDescriptor.new(0, 'in_RGBA', ll.PortDirection.In, ll.PortType.ImageView))
    desc:addPort(ll.PortDescriptor.new(1, 'out_RGBA', ll.PortDirection.Out, ll.PortType.ImageView))

    -- parameter with default value
    desc:setParameter('levels', 2)

    return desc
end


function builder.onNodeInit(node)

    levels = node.descriptor:getParameter('levels')
    ll.logd('ImagePyramid', 'onNodeInit: levels:', levels)

    -- in_RGBA should have been bound before calling init()
    in_RGBA = node:getPort('in_RGBA')
    node:bind(string.format('out_RGBA_downY_0', i), in_RGBA)

    for i = 1, levels -1 do
        ll.logd('ImagePyramid', 'onNodeInit: level:', i)
        
        downX = ll.createComputeNode('ImageDownsampleX')
        downY = ll.createComputeNode('ImageDownsampleY')

        downX:bind('in_RGBA', in_RGBA)
        downX:init()

        downY:bind('in_RGBA', downX:getPort('out_RGBA'))
        downY:init()

        in_RGBA = downY:getPort('out_RGBA')

        -- bind the output
        if i == levels -1 then
            node:bind('out_RGBA', downY:getPort('out_RGBA'))
        end

        node:bindNode(string.format('downX_%d', i), downX)
        node:bindNode(string.format('downY_%d', i), downY)

        node:bind(string.format('out_RGBA_downX_%d', i), downX:getPort('out_RGBA'))
        node:bind(string.format('out_RGBA_downY_%d', i), downY:getPort('out_RGBA'))
    end

end


function builder.onNodeRecord(node, cmdBuffer)

    levels = node.descriptor:getParameter('levels')

    for i = 1, levels -1 do
        ll.logd('ImagePyramid', 'onNodeRecord: level:', i)

        downX = node:getNode(string.format('downX_%d', i))
        downY = node:getNode(string.format('downY_%d', i))

        downX:record(cmdBuffer)
        cmdBuffer:memoryBarrier()
        downY:record(cmdBuffer)
        cmdBuffer:memoryBarrier()
    end
end


ll.registerNodeBuilder('ImagePyramid', builder)
